## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [21]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [22]:
# View the column names in the data
df.columns
print(df.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [23]:
# Use the describe function to gather some basic statistics
df.describe()
statistics = df.describe().round(2)
print(statistics)


       client_id    order_id  order_week  order_year  unit_price  unit_cost  \
count   54639.00    54639.00    54639.00    54639.00    54639.00   54639.00   
mean    54837.87  5470190.03       11.36     2022.99      136.27      99.45   
std     25487.44  2599806.93        7.02        0.08      183.87     133.16   
min     10033.00  1000886.00        1.00     2022.00        0.01       0.01   
25%     33593.00  3196372.00        6.00     2023.00       20.80      14.84   
50%     53305.00  5496966.00       11.00     2023.00       68.31      49.89   
75%     78498.00  7733869.00       17.00     2023.00      173.16     125.57   
max     99984.00  9998480.00       52.00     2023.00     1396.23     846.27   

       unit_weight         qty  line_number  
count     54639.00    54639.00     54639.00  
mean          5.00      570.26         2.98  
std           5.33    18795.52         2.44  
min           0.00        0.00         0.00  
25%           1.45       32.00         1.00  
50%         

In [24]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()
df.count()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

first          54639
last           54639
job            54639
phone          54639
email          54639
client_id      54639
order_id       54639
order_date     54639
order_week     54639
order_year     54639
item_id        54639
category       54639
subcategory    54639
unit_price     54639
unit_cost      54639
unit_weight    54639
qty            54639
line_number    54639
dtype: int64

In [25]:
# What three item categories had the most entries?
df['category']= df['category'].astype(str)
top_three_categories = df['category'].value_counts().head(3)
print(top_three_categories)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [26]:
# For the category with the most entries, which subcategory had the most entries?
most_common_category = df['category'].value_counts().idxmax()
filtered_df = df[df['category'] == most_common_category]
subcategory_counts = filtered_df['subcategory'].value_counts()
print(subcategory_counts)
top_subcategory = subcategory_counts.idxmax()
print(top_subcategory, "has the most subcategory entries within consumables") 

subcategory
bathroom supplies    6424
paper products       5874
kitchen supplies     5746
misc                 2891
pens                 2603
Name: count, dtype: int64
bathroom supplies has the most subcategory entries within consumables


In [27]:
# Which five clients had the most entries in the data?
top_five_clients = df['client_id'].value_counts()
top_five_clients = top_five_clients.head(5)
print(top_five_clients)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [28]:
# Store the client ids of those top 5 clients in a list.
top_five_clients_ids = top_five_clients.index.tolist() 
print("Top 5 clients by quantity:") 
print(top_five_clients_ids)

Top 5 clients by quantity:
[33615, 66037, 46820, 38378, 24741]


In [29]:
# How many total units (the qty column) did the client with the most entries order?
client_id_qty = 33615
client_id_qty_total = df.loc[df['client_id'] == client_id_qty, 'qty'].sum()
print(client_id_qty_total)


64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [30]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
print(df['subtotal'].head())

0    115164.00
1       523.95
2       527.28
3      1056.18
4      3902.00
Name: subtotal, dtype: float64


In [31]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def calculation_shipping_price(row):
    item_weight = row['unit_weight']
    if item_weight > 50:
        shipping_price = item_weight * 7
    else: 
        shipping_price = item_weight * 10
    return shipping_price
df['shipping_price'] = df.apply(calculation_shipping_price, axis=1) 
print(df[['unit_weight', 'shipping_price']].head())

   unit_weight  shipping_price
0         7.50            75.0
1         1.49            14.9
2         1.68            16.8
3         1.23            12.3
4        46.43           464.3


In [32]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
sales_tax_rate = 0.0925
df['total_price'] = df['subtotal'] + df['shipping_price'] + df['subtotal'] * sales_tax_rate 
print(df['total_price'])



0        125891.670000
1           587.315375
2           592.853400
3          1166.176650
4          4727.235000
             ...      
54634      3019.544325
54635     10724.880025
54636     34115.036250
54637       361.492800
54638      3411.066175
Name: total_price, Length: 54639, dtype: float64


In [33]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price'] 
print(df[['unit_cost', 'qty', 'shipping_price', 'line_cost']].head())


   unit_cost  qty  shipping_price  line_cost
0     762.71  105            75.0   80159.55
1      15.09   21            14.9     331.79
2       7.86   39            16.8     323.34
3      24.85   29            12.3     732.95
4     108.17   20           464.3    2627.70


In [34]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['total_price'] - df['line_cost'] 
print(df[['total_price', 'line_cost', 'line_profit']].head())


     total_price  line_cost   line_profit
0  125891.670000   80159.55  45732.120000
1     587.315375     331.79    255.525375
2     592.853400     323.34    269.513400
3    1166.176650     732.95    433.226650
4    4727.235000    2627.70   2099.535000


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [35]:
# Check your work using the totals above
order1_total_price = df[df['order_id'] == 2742071]['total_price'].sum().round(2)
order2_total_price = df[df['order_id'] == 2173913]['total_price'].sum().round(2)
order3_total_price = df[df['order_id'] == 6128929]['total_price'].sum().round(2)
print("Total Price for Order ID 2742071:", order1_total_price)
print("Total Price for Order ID 2173913:", order2_total_price) 
print("Total Price for Order ID 6128929:", order3_total_price)

Total Price for Order ID 2742071: 139825.57
Total Price for Order ID 2173913: 119815.13
Total Price for Order ID 6128929: 752707.06


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [36]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_five_clients_spending = df[df['client_id'].isin(top_five_clients_ids)].groupby('client_id')['total_price'].sum()
print("Total spending for the top 5 clients:")
print(top_five_clients_spending)


Total spending for the top 5 clients:
client_id
24741    7.667761e+07
33615    6.391752e+06
38378    9.171766e+06
46820    8.005306e+06
66037    8.745329e+06
Name: total_price, dtype: float64


In [37]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_five_totals = df[df['client_id'].isin(top_five_clients_ids)].groupby('client_id').agg( 
    total_units_purchased=('qty', 'sum'), 
    total_shipping_price=('shipping_price', 'sum'), 
    total_revenue=('total_price', 'sum'), 
    total_profit=('line_profit', 'sum') 
    )
print("Summary DataFrame for Top 5 Clients:") 
print(top_five_totals)


Summary DataFrame for Top 5 Clients:
           total_units_purchased  total_shipping_price  total_revenue  \
client_id                                                               
24741                     239862                9365.6   7.667761e+07   
33615                      64313               12609.4   6.391752e+06   
38378                      73667               11895.0   9.171766e+06   
46820                      75768               11094.8   8.005306e+06   
66037                      43018               10017.3   8.745329e+06   

           total_profit  
client_id                
24741      3.610580e+07  
33615      2.032814e+06  
38378      2.954605e+06  
46820      2.588469e+06  
66037      3.125980e+06  


In [38]:
# Format the data and rename the columns to names suitable for presentation.
top_five_clients_totals_data = pd.DataFrame(top_five_totals)

# Define the money columns. 
money_columns = ('total_shipping_price', 'total_revenue', 'total_profit')

# Define a function that converts a dollar amount to millions.
def convert_to_millions(amount):
    if abs(amount) >= 1_000_000:
        return '${:,.2f}M'.format(amount / 1_000_000)
    else:
        return '${:,.2f}'.format(amount)

# Apply the currency_format_millions function to only the money columns. 
for col in money_columns:
    top_five_clients_totals_data[col] = top_five_clients_totals_data[col].apply(convert_to_millions)

# Rename the columns to reflect the change in the money format. 
renamed_columns = top_five_clients_totals_data.rename(columns={
    'total_units_purchased': 'Total Units Purchased (Millions)',
    'total_shipping_price': 'Total Shipping Price (Millions)',
    'total_revenue': 'Total Revenue (Millions)',
    'total_profit': 'Total Profit (Millions)'})
print(renamed_columns)

           Total Units Purchased (Millions) Total Shipping Price (Millions)  \
client_id                                                                     
24741                                239862                       $9,365.60   
33615                                 64313                      $12,609.40   
38378                                 73667                      $11,895.00   
46820                                 75768                      $11,094.80   
66037                                 43018                      $10,017.30   

          Total Revenue (Millions) Total Profit (Millions)  
client_id                                                   
24741                      $76.68M                 $36.11M  
33615                       $6.39M                  $2.03M  
38378                       $9.17M                  $2.95M  
46820                       $8.01M                  $2.59M  
66037                       $8.75M                  $3.13M  


In [39]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_data = renamed_columns.sort_values(by='Total Profit (Millions)', ascending=False)



In [40]:
# Print the sorted DataFrame
print(sorted_data)

           Total Units Purchased (Millions) Total Shipping Price (Millions)  \
client_id                                                                     
24741                                239862                       $9,365.60   
66037                                 43018                      $10,017.30   
38378                                 73667                      $11,895.00   
46820                                 75768                      $11,094.80   
33615                                 64313                      $12,609.40   

          Total Revenue (Millions) Total Profit (Millions)  
client_id                                                   
24741                      $76.68M                 $36.11M  
66037                       $8.75M                  $3.13M  
38378                       $9.17M                  $2.95M  
46820                       $8.01M                  $2.59M  
33615                       $6.39M                  $2.03M  
